### PR TITLE
Add DNS Entry Checker

### PR DESCRIPTION
**Pull request by charlieDurnsfordHollands**
Adds a script that checks for missing and mismatching IPs and their DNS' in order to more easily identify and investigate them.

### CODE REVIEW

File: .github/workflows/dns_entry_checker.yaml [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+  push:
+    branches:
+      - master
+  pull_request:
+    paths:
+      - "dns_entry_checker/**"
+      - ".github/workflows/dns_entry_checker.yaml"
+
+jobs:
+  test_and_lint:
```
Comment (anish-mudaraddi) [inline .github/workflows/dns_entry_checker.yaml:13]: As mentioned in #77 you're not actually doing any linting 

```suggestion
  test_with_unittest:
```
You can add this to make github actions perform linting

```
    - name: Analyse with pylint 
      run: |
          pylint --recursive yes test dns_entry_checker
```
you'll also need to put a .pylintrc file in dns_entry_checker folder

File: dns_entry_checker/dns_entry_checker.py [hunk 2]
```diff
@@ -0,0 +1,127 @@
+import paramiko
+from re import compile
+from getpass import getpass
+from collections import defaultdict
+
+
+# Creates an SSH client through Paramiko with the users details and returns the client
+def create_client(host, user, password):
```
Comment (anish-mudaraddi) [inline dns_entry_checker/dns_entry_checker.py:8]: Looks like you have some duplicate code here in #77 - this hints to me that perhaps we should combine your two scripts into a single package.

I'd say it's fine for now - but its worth looking at in the future.

Comment (anish-mudaraddi) [inline dns_entry_checker/dns_entry_checker.py:8]: Please Add Docstrings to these functions too - same as #77

File: dns_entry_checker/dns_entry_checker.py [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+        if returned_dns:
+            backward_ip = ip_rexp.findall(returned_dns[0])
+            if not backward_ip:
+                backward_missing_file.write(f"{ips[1]}\n")
+            elif not backward_ip[0][0] == ips[0]:
+                backward_mismatch_file.write(f"{ips[1]}\n")
+        else:
+            backward_missing_file.write(f"{ips[1]}\n")
+    else:
+        forward_mismatch_file.write(f"{ips[1]}\n")
```
Comment (anish-mudaraddi) [inline dns_entry_checker/dns_entry_checker.py:54]: try to avoid nested ifs. 
As a rule of thumb - if you see something with 3 or more levels of indentation - it's a code smell - since its hard to follow logically 

```
    if not ips[0].replace("-", ".") == ips[1]:
        forward_mismatch_file.write(f"{ips[1]}\n")
        return 
    returned_dns = ssh_command(client, "dig -x {s} +short".format(s=ips[1]))
    if not returned_dns
        backward_missing_file.write(f"{ips[1]}\n")
        return
    backward_ip = ip_rexp.findall(returned_dns[0])
    if not backward_ip:
        backward_missing_file.write(f"{ips[1]}\n")
    elif not backward_ip[0][0] == ips[0]:
        backward_mismatch_file.write(f"{ips[1]}\n")
```

File: dns_entry_checker/dns_entry_checker.py [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+    order_values = last_rexp.findall(ips[1])[0].split(".")
+    order_check_dict[order_values[0]].append(order_values[1])
+    return ips
+
+
+# Checks if the DNS returns the matching IP
+def check_ip_dns_mismatch(
+    ips,
+    client,
+    ip_rexp,
```
Comment (anish-mudaraddi) [inline dns_entry_checker/dns_entry_checker.py:38]: Since this variable is immutable (doesn't change)
you can set `ip_rexp` as a global variable at the top of the script
`IP_REXP = compile(r"(([0-9]{1,3}[.-]){3}[0-9]{1,3})")`

That way you don't have to use it here as a parameter

File: dns_entry_checker/dns_entry_checker.py [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+        else:
+            backward_missing_file.write(f"{ips[1]}\n")
+    else:
+        forward_mismatch_file.write(f"{ips[1]}\n")
+
+
+# Checks which IPs are missing from 2-254
+def check_missing_ips(key, gap_missing_file, i):
+    res = set([eval(j) for j in key[1]])
+    missing_values = [k for k in sorted(set(range(2, 255))) if k not in res]
```
Comment (anish-mudaraddi) [inline dns_entry_checker/dns_entry_checker.py:60]: why are you doing sorted(set(range(2, 255)))?
surely this works fine
```suggestion
    missing_values = [k for k in range(2, 255) if k not in res]
```

File: dns_entry_checker/dns_entry_checker.py [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+    for ip in missing_values:
+        gap_missing_file.write(f"172.16.{i + 100}.{ip}\n")
+
+
+def dns_entry_checker():
+    # Prompt the user for their FedID, Password and IP to connect to
+    user = input("FedID: ")
+    password = getpass("Password: ")
+    ip = input("IP: ")
```
Comment (anish-mudaraddi) [inline dns_entry_checker/dns_entry_checker.py:69]: see the comment on `argparse` in #77

File: dns_entry_checker/dns_entry_checker.py [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+
+    # Create or clear, then open to append to, the output files
+    open("output\\forward_mismatch_list.txt", "w").close()
+    open("output\\backward_mismatch_list.txt", "w").close()
+    open("output\\backward_missing_list.txt", "w").close()
+    open("output\\gap_missing_list.txt", "w").close()
+    forward_mismatch_file = open("output\\forward_mismatch_list.txt", "a")
+    backward_mismatch_file = open("output\\backward_mismatch_list.txt", "a")
+    backward_missing_file = open("output\\backward_missing_list.txt", "a")
+    gap_missing_file = open("output\\gap_missing_list.txt", "a")
```
Comment (anish-mudaraddi) [inline dns_entry_checker/dns_entry_checker.py:91]: see the comment about not using open and close for files in #77

File: dns_entry_checker/dns_entry_checker.py [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+
+    # Define the Regex used to identify IPs
+    ip_rexp = compile(r"(([0-9]{1,3}[.-]){3}[0-9]{1,3})")
+
+    # Create a dictionary to store IPs for checking gaps
+    order_check_dict = defaultdict(list)
```
Comment (anish-mudaraddi) [inline dns_entry_checker/dns_entry_checker.py:97]: why are you using defaultdict here and not just dict?

File: dns_entry_checker/dns_entry_checker.py [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+# Checks if the IP address and DNS match
+def pair_ip_and_dns(dns_pair, order_check_dict, ip_rexp):
+    last_rexp = compile(r"([0-9]{1,3}\.[0-9]{1,3}(?!.))")
+
+    ips = ip_rexp.findall(dns_pair)
+    ips = [ips[0][0], ips[1][0]]
+
+    order_values = last_rexp.findall(ips[1])[0].split(".")
+    order_check_dict[order_values[0]].append(order_values[1])
+    return ips
```
Comment (anish-mudaraddi) [inline dns_entry_checker/dns_entry_checker.py:31]: This function is doing 2 things - setting ips and populating order_values - this should be split into two separate functions which are called sequentially - one to get `ips` and another to populate `order_check_dict`

**Review by anish-mudaraddi — CHANGES_REQUESTED** [review]

File: dns_entry_checker/dns_entry_checker.py (hunk 8 above)
Comment (charlieDurnsfordHollands) [reply dns_entry_checker/dns_entry_checker.py:197]: Using a defaultdict means that if the list is appended to for a key that isn't in the dictionary yet, it automatically adds the key. Using a normal dict would work if line 57 is changed to:
```
    if order_values[0] not in order_check_dict.keys():
        order_check_dict[order_values[0]] = list(order_values[1])
    else:
        order_check_dict[order_values[0]].append(order_values[1])
```
instead of just:
```
order_check_dict[order_values[0]].append(order_values[1])
```
If the version using the dict is preferable then I'm fine with changing it though.

Comment (anish-mudaraddi) [reply dns_entry_checker/dns_entry_checker.py:197]: oh cool - did not know this - you can keep it as is.

File: dns_entry_checker/dns_entry_checker.py [hunk 10]
```diff
@@ -0,0 +1,223 @@
...
+        output or "output", "forward_mismatch_list.txt"
+    )
+    backward_mismatch_filepath = os.path.join(
+        output or "output", "backward_mismatch_list.txt"
+    )
+    backward_missing_filepath = os.path.join(
+        output or "output", "backward_missing_list.txt"
+    )
+    gap_missing_filepath = os.path.join(output or "output", "gap_missing_list.txt")
+    check_output_files(
```
Comment (anish-mudaraddi) [inline dns_entry_checker/dns_entry_checker.py:189]: you can rewrite check_output_files function to be more generic and call it multiple times for each file
something like 

```
for filepath in [forward_mismatch_filepath, backward_mismatch_filepath, backward_missing_filepath, gap_missing_filepath]:
    check_file(filepath)
```

File: dns_entry_checker/dns_entry_checker.py [hunk 11]
```diff
@@ -0,0 +1,223 @@
...
+        metavar="OUTPUT",
+        help="Directory to create the output files in",
+    )
+
+    # Define the variables with the script arguments
+    args = parser.parse_args()
+    user = args.user
+    password = args.password
+    ip = args.ip
+    output = args.output
```
Comment (anish-mudaraddi) [inline dns_entry_checker/dns_entry_checker.py:164]: you can move this into another function - and return a dictionary of argument names : argument values 
```
def parse_args(inp_args):
    """
    Function to parse commandline args
    :param inp_args: a set of commandline args to parse (dict)
    :returns: A dictionary of parsed args 
    """
    ...code here..
    args = parser.parse_args(inp_args)
    return args
```
you can call the function like so
```
args = parse_args(sys.argv[1:])

File: dns_entry_checker/test/test_dns_entry_checker.py [hunk 12]
```diff
@@ -0,0 +1,193 @@
...
+class DNSEntryCheckerTests(unittest.TestCase):
+    @patch("dns_entry_checker.paramiko")
+    def test_create_client(self, mock_paramiko):
+        host = "test.aquilon.gridpp.rl.ac.uk"
+        user = "test.abc12345"
+        password = "test.A4gdfGs"
+
+        mock_client = MagicMock()
+        mock_paramiko.client.SSHClient.return_value = mock_client
+        mock_paramiko.client.set_missing_host_key_policy = MagicMock()
```
Comment (anish-mudaraddi) [inline dns_entry_checker/test/test_dns_entry_checker.py:24]: You don't need this line - since mock_paramiko will automatically set it to a magic mock
```suggestion
```

File: dns_entry_checker/test/test_dns_entry_checker.py [hunk 13]
```diff
@@ -0,0 +1,193 @@
...
+        mock_paramiko.AutoAddPolicy = NonCallableMock()
+        mock_paramiko.connect = MagicMock()
+
+        client = create_client(host, user, password)
+
+        client.connect.assert_called_once_with(
```
Comment (anish-mudaraddi) [inline dns_entry_checker/test/test_dns_entry_checker.py:30]: Add this line to test that set_missing_host_key_policy function was called property
```suggestion
        client.set_missing_host_key_policy.assert_called_once_with(mock_paramiko.AutoAddPolicy)
        client.connect.assert_called_once_with(
```

File: dns_entry_checker/test/test_dns_entry_checker.py [hunk 14]
```diff
@@ -0,0 +1,193 @@
...
+
+        mock_client.exec_command.assert_called_once_with(
+            command, get_pty=True, timeout=60
+        )
+        mock_stdout.channel.recv_exit_status.assert_called_once()
+        mock_stdout.readlines.assert_called_once()
+
+        self.assertEqual(command_output, mock_stdout.readlines.return_value)
+
+    def test_pair_ip_and_dns(self):
```
Comment (anish-mudaraddi) [inline dns_entry_checker/test/test_dns_entry_checker.py:55]: rename this test method so its more descriptive
```suggestion
    def test_find_ip_dns_pair_valid(self):
```

File: dns_entry_checker/test/test_dns_entry_checker.py [hunk 15]
```diff
@@ -0,0 +1,193 @@
...
+    )
+    @mock.patch("builtins.open")
+    def test_check_ip_dns_mismatch(self, name, returned_ips, expected_out, mock_file):
+        ips_dns_pair = returned_ips
+        client = MagicMock()
+        backward_mismatch_filepath = "test/backward/mismatch/filepath"
+        forward_mismatch_filepath = "test/forward/mismatch/filepath"
+        backward_missing_filepath = "test/backward/missing/filepath"
+
+        with patch("dns_entry_checker.ssh_command"):
```
Comment (anish-mudaraddi) [inline dns_entry_checker/test/test_dns_entry_checker.py:75]: why are you doing `with patch` here instead of `mock.patch`

File: dns_entry_checker/test/test_dns_entry_checker.py [hunk 16]
```diff
@@ -0,0 +1,193 @@
...
+        check_missing_ips(key, gap_missing_filepath)
+
+        print(expected_out, mock_file.mock_calls)
+
+        if expected_out:
+            assert call("test/gap/missing/filepath", "a") in mock_file.mock_calls
+        else:
+            assert call("test/gap/missing/filepath", "a") not in mock_file.mock_calls
+
+
```
Comment (anish-mudaraddi) [inline dns_entry_checker/test/test_dns_entry_checker.py:191]: you're missing tests for check_output_files and populate_ip_dict

i'd recommend using `coverage` python tool to see if you covered all tests
https://www.pythontutorial.net/python-unit-testing/python-unittest-coverage/

**Review by anish-mudaraddi — CHANGES_REQUESTED** [review]
left a few more comments

**Review by anish-mudaraddi — APPROVED** [review]
Okay this looks good - if you mark it ready for review we can merge it
